sys/uri_parser: minimal and non-destructive URI parsing

cgundogan opened the pull request:
### Contribution description
Some places in the code base re-implement (e.g., waakama, sock_util) parsing of URIs. I need a URI parser for a project and implemented a minimal and non-destructive parser as a module under `/sys`. It is non-destructive, because the passed URI string is not modified.

After parsing a NULL-termianted URI string, the parts are accessible through a `uri_parser_result_t` struct. 

As follow-up, we could add iterators for the separate query fields and improve the unittests once we have more features.

### Testing procedure
A `unittest` is attached.
`make -C tests/unittests tests-uri_parser term`

### Issues/PRs references
none

bergzand commented:
Should the code here be safe for usage with untrusted input (e.g. strings received from coap requests such as used in the SUIT module)?

cgundogan commented:
In general yes, this would require a more elaborate testing suite to be on the safe side, though.

benpicco commented:
That would be another use-case for #13157 :wink:

cgundogan commented on sys/uri_parser/uri_parser.c:
```diff
@@ -0,0 +1,178 @@
+    /* first character should be ALPHA */
+    if (!(((uri[0] >= 'A') && (uri[0] <= 'Z')) ||
+          ((uri[0] >= 'a') && (uri[0] <= 'z')))) {
+        NULL;
```
oh man, `return NULL` (:

bergzand commented:
I wonder if the API should include the string length:

` int uri_parser_process(uri_parser_result_t *result, const char *uri, size_t uri_len)`

This way we don't have to trust the source to end the string with a `\0`

> That would be another use-case for #13157 wink

Working on this.

bergzand commented on tests/unittests/tests-uri_parser/tests-uri_parser.c:
```diff
@@ -0,0 +1,103 @@
+    uri_parser_result_t ures;
+    int res = -1;
+
+    memset(&ures, 0, sizeof(ures));
```
Should this memset be moved to `uri_parser_process`?

bergzand commented on sys/uri_parser/uri_parser.c:
```diff
@@ -0,0 +1,178 @@
+
+    /* check for port after host part */
+    /* repeat until last ':' because of multiple occurrences in IPv6 addresses */
+    char *port_begin, *p = result->host;
```
```suggestion
    char *port_begin = NULL;
    char *p = result->host;
```

bergzand commented:
AFL detects crashes on the following strings:

```
coap://R@[2001:db8::1]:5own/v=1
coap://R@[2001:db8::1]:5own/:v=1
cap://R@[2001:db8::1]:5own/?v=1
oap://Y2001:db8::1]:5own/av=1
//Rb[ʰ00J:d/5v=0
coap://oap://P@[2001:b
coap:///R@[2008::1]:5own//R@[2008::1]:5own/?v=1
coaP://R/RZ[2001[8:01[8::1]:5o:1]:5oTMv=1
coap://R@////////////////7///v=1
coa[:////[2001:db5ow:5own/Ov=1
```

Have fun! :wink:

Achilleshiel commented on sys/uri_parser/uri_parser.c:
```diff
@@ -0,0 +1,178 @@
+        result->path_len -= result->query_len + 1;
+    }
+
+    return (result->query + result->query_len);
```
If the URI does not contain a query, the if-statement is not executed. Assuming that `result` is not initialized with zeros, the return-statement segfaults, as the charptr `query` and `query_len` point to a random location.
```suggestion
    char *path_end = strchr(uri, '?');

    if (path_end) {
        result->query = path_end + 1;
        /* do not count '?' */
        result->query_len = result->path_len - (path_end - uri) - 1;
        /* cut path part before query and '?' */
        result->path_len -= result->query_len + 1;
        
        return (result->query + result->query_len);
    }

    return NULL;
```

Achilleshiel replied:
If the URI does not contain a query, the if-statement is not executed. Assuming that `result` is not initialized with zeros, the return-statement segfaults, as the charptr `query` and `query_len` point to a random location.
```suggestion
        return (result->query + result->query_len);
    }

    return NULL;
```

Achilleshiel commented on sys/uri_parser/uri_parser.c:
```diff
@@ -0,0 +1,178 @@
+        return -1;
+    }
+
+    return 0;
```
This looks similar to the `_parse_relative` function.
```suggestion
    return _parse_relative(result, uri);
```

Achilleshiel commented on sys/uri_parser/uri_parser.c:
```diff
@@ -0,0 +1,178 @@
+    /* check for query start '?' */
+    char *path_end = strchr(uri, '?');
+
+    if (path_end) {
```
If the URI does not contain a query, the if-statement is not executed. Assuming that `result` is not initialized with zeros, the return-statement segfaults, as the charptr `query` and `query_len` point to a random location.
```suggestion
    if (!path_end) {
        return NULL;
    }
    else {
```

bergzand commented:
Another test vector:
```
tel:+1-816-555-1212
```

cgundogan commented:
thanks @Achilleshiel  @bergzand for this extensive testing! I did a couple of fixups. It still fails for the `tel:...` vector :/

cgundogan commented:
> AFL detects crashes on the following strings:
> 
> ```
> coap://R@[2001:db8::1]:5own/v=1
> coap://R@[2001:db8::1]:5own/:v=1
> cap://R@[2001:db8::1]:5own/?v=1
> oap://Y2001:db8::1]:5own/av=1
> //Rb[ʰ00J:d/5v=0
> coap://oap://P@[2001:b
> coap:///R@[2008::1]:5own//R@[2008::1]:5own/?v=1
> coaP://R/RZ[2001[8:01[8::1]:5o:1]:5oTMv=1
> coap://R@////////////////7///v=1
> coa[:////[2001:db5ow:5own/Ov=1
> ```
> 
> Have fun!

these vectors do not crash anymore, but I do not yet know if everything was correctly parsed into the result container.

silkeh commented:
Perhaps it would be worthwhile to add test vectors from [Wikipedia's List of URI schemes](https://en.wikipedia.org/wiki/List_of_URI_schemes), like `sms:+15105550101,+15105550102?body=hello%20there`

bergzand commented:
And another test vector for crashes:
```
a
```

cgundogan commented:
thanks @Achilleshiel @bergzand @silkeh for the feedback. I added [further test vectors](https://github.com/RIOT-OS/RIOT/pull/13758/files#diff-e1a2b9c48971d8d2340c1a0fd715ebf5R53) to the unittests and included a few parsing fixes.

cgundogan commented:
No idea why codespell isn't able to complete in Travis, but I addressed all other issues that Travis reported (besides the squashung).

cgundogan commented:
@bergzand I modified the parsing to work on non-NULL-terminated strings. Then I added a parse function `uri_parser_process_string()` that calls the proper function with `strlen(uri)` as argument.

miri64 commented on tests/unittests/tests-uri_parser/tests-uri_parser.c:
```diff
@@ -0,0 +1,293 @@
+      scheme, userinfo, host, port,
+      path, query, expected return value)
+*/
+const validate_t validate_uris[24] = {
```
```suggestion
static const validate_t validate_uris[24] = {
```

?

cgundogan replied:
yup, amended the small change. thanks :)

miri64 commented on tests/unittests/tests-uri_parser/tests-uri_parser.c:
```diff
@@ -0,0 +1,293 @@
+        "",
+        "oasis:names:specification:docbook:dtd:xml:4.1.2",
+        "",
+        0),
```
At least one more test vector should be added here:

```C
    VEC("",
        "",
        "",
        "",
        "",
        "",
        "",
        -1),
```

miri64 commented on sys/include/uri_parser.h:
```diff
@@ -0,0 +1,101 @@
+ * @brief Parse a URI
+ *
+ * @param[out]  result    pointer to a container that will hold the result
+ * @param[in]   uri       NULL-terminated URI to parse
```
```suggestion
 * @param[in]   uri       Null-terminated URI to parse
```

or better

```suggestion
 * @param[in]   uri       Zero-terminated URI to parse
```

`NULL` is a constant representing a zero-valued `void` pointer ;-).

miri64 replied:
I tried also this test vector:

```C
TEST_ASSERT_EQUAL_INT(-1, uri_parser_process_string(&ures, NULL));
```

Which led to a crash. I think this should be documented:
```suggestion
 * @param[in]   uri       Zero-terminated URI to parse. Must not be `NULL`
```

and a precondition should be added

```
@pre `uri != NULL`
```

For the other functions these doc amendments should probably be added as well as well!

miri64 commented:
Also a weird test case where I don't even understand what is happening: `"/"`?

miri64 commented on tests/unittests/tests-uri_parser/tests-uri_parser.c:
```diff
@@ -0,0 +1,293 @@
+        else {                                                          \
+            TEST_ASSERT_EQUAL_INT(strlen(validate_uris[i].comp),        \
+                                  ures.comp##_len);                     \
+            TEST_ASSERT_EQUAL_INT(0,                                    \
+                                  memcmp(ures.comp,                     \
+                                         validate_uris[i].comp,         \
+                                         strlen(validate_uris[i].comp))); \
```
Ahhhh this is what was confusing me. Can we use maybe

```suggestion
            TEST_ASSERT_EQUAL_STRING(ures.comp,                         \
                                     validate_uris[i].comp);            \
```

here?

miri64 replied:
I tried it myself. Since the strings in `ures` are not zero-terminated: no.

cgundogan replied:
Unfortunately this doesn't work, because `ures.comp` is not `Zero`-terminated. Since `TEST_ASSERT_EQUAL_STRING` is using `strcmp()` underneath, it just walks over the valid memory region .. since there is no strncmp test macro, I went with memcmp.

miri64 replied:
Why `memcmp` and not `strncmp`?
